substrate/runtime/primitives/src/generic.rs

```diff
@@ -65,53 +65,79 @@ impl<AccountId, Index, Call> Slicable for Extrinsic<AccountId, Index, Call> wher
+/// A extrinsic right from the external world. Unchecked.
 #[derive(PartialEq, Eq, Clone)]
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
-pub struct UncheckedExtrinsic<AccountId, Index, Call, Signature> {
+pub struct UncheckedExtrinsic<Address, Index, Call, Signature> {
 	/// The actual extrinsic information.
-	pub extrinsic: Extrinsic<AccountId, Index, Call>,
-	/// The signature; should be an Ed25519 signature applied to the serialised `extrinsic` field.
+	pub extrinsic: Extrinsic<Address, Index, Call>,
+	/// The signature.
 	pub signature: Signature,
 }
 
-impl<AccountId, Index, Call, Signature> traits::Checkable for UncheckedExtrinsic<AccountId, Index, Call, Signature> where
- 	AccountId: Member + MaybeDisplay,
- 	Index: Member + MaybeDisplay + SimpleArithmetic,
- 	Call: Member,
-	Signature: Member + traits::Verify<Signer = AccountId>,
-	Extrinsic<AccountId, Index, Call>: Slicable,
-{
-	type Checked = CheckedExtrinsic<AccountId, Index, Call, Signature>;
-
-	fn check(self) -> Result<Self::Checked, Self> {
-		if ::verify_encoded_lazy(&self.signature, &self.extrinsic, &self.extrinsic.signed) {
-			Ok(CheckedExtrinsic(self))
-		} else {
-			Err(self)
+impl<Address, Index, Call, Signature> UncheckedExtrinsic<Address, Index, Call, Signature> {
+	/// New instance.
+	pub fn new(extrinsic: Extrinsic<Address, Index, Call>, signature: Signature) -> Self {
+		UncheckedExtrinsic {
+			extrinsic,
+			signature,
 		}
 	}
 }
 
-impl<AccountId, Index, Call, Signature> UncheckedExtrinsic<AccountId, Index, Call, ::MaybeUnsigned<Signature>> where
- 	AccountId: Member + Default + MaybeDisplay,
- 	Index: Member + MaybeDisplay + SimpleArithmetic,
- 	Call: Member,
-	Signature: Member + Default + traits::Verify<Signer = AccountId>,
-	Extrinsic<AccountId, Index, Call>: Slicable,
+impl<Address, AccountId, Index, Call, Signature> UncheckedExtrinsic<Address, Index, Call, ::MaybeUnsigned<Signature>> where
+	Signature: traits::Verify<Signer=AccountId> + Default + Eq,
+	AccountId: Default + Eq,
 {
-	/// Is this extrinsic signed?
+	/// `true` if this extrinsic is signed.
 	pub fn is_signed(&self) -> bool {
-		self.signature.is_signed(&self.extrinsic.signed)
+		self.signature.is_signed()
 	}
 }
 
-impl<AccountId, Index, Call, Signature> Slicable for UncheckedExtrinsic<AccountId, Index, Call, Signature> where
- 	AccountId: Member + MaybeDisplay,
- 	Index: Member + MaybeDisplay + SimpleArithmetic,
- 	Call: Member,
-	Signature: Member + Slicable,
+impl<Address, AccountId, Index, Call, Signature> traits::Checkable
+	for UncheckedExtrinsic<Address, Index, Call, ::MaybeUnsigned<Signature>>
+where
+	Address: Member + Default + MaybeDisplay,
+	Index: Member + MaybeDisplay + SimpleArithmetic,
+	Call: Member,
+	Signature: traits::Verify<Signer=AccountId> + Eq + Default,
+	AccountId: Member + Default + MaybeDisplay,
+	::MaybeUnsigned<Signature>: Member,
 	Extrinsic<AccountId, Index, Call>: Slicable,
+{
+	type Address = Address;
+	type AccountId = AccountId;
```

Comment:
The approach here is a little too restrictive on both the `Checkable` trait and extrinsic type. For example, how would we define the checkable trait for a normal extrinsic that uses an secp256k1 signature? But this is good enough for the current PR and we can refactor further a bit later.